src/account_manager.rs

```diff
@@ -259,62 +328,74 @@ impl AccountManager {
-        })
+    pub async fn get_accounts(&self) -> Vec<AccountHandle> {
+        let accounts = self.accounts.read().await;
+        accounts.values().cloned().collect()
     }
 
     /// Reattaches an unconfirmed transaction.
     pub async fn reattach(&self, account_id: &AccountIdentifier, message_id: &MessageId) -> crate::Result<Message> {
-        let mut account = self.get_account(account_id)?;
-        account.sync().execute().await?.reattach(message_id).await
+        let account = self.get_account(account_id).await?;
+        account.sync().await.execute().await?.reattach(message_id).await
     }
 
     /// Promotes an unconfirmed transaction.
     pub async fn promote(&self, account_id: &AccountIdentifier, message_id: &MessageId) -> crate::Result<Message> {
-        let mut account = self.get_account(account_id)?;
-        account.sync().execute().await?.promote(message_id).await
+        let account = self.get_account(account_id).await?;
+        account.sync().await.execute().await?.promote(message_id).await
     }
 
     /// Retries an unconfirmed transaction.
     pub async fn retry(&self, account_id: &AccountIdentifier, message_id: &MessageId) -> crate::Result<Message> {
-        let mut account = self.get_account(account_id)?;
-        account.sync().execute().await?.retry(message_id).await
+        let account = self.get_account(account_id).await?;
+        account.sync().await.execute().await?.retry(message_id).await
     }
 }
 
-async fn poll(storage_path: PathBuf, syncing: bool) -> crate::Result<()> {
+async fn poll(accounts: AccountStore, storage_path: PathBuf, syncing: bool) -> crate::Result<()> {
     let retried = if syncing {
-        let accounts_before_sync = crate::storage::with_adapter(&storage_path, |storage| storage.get_all())?;
-        let mut accounts_before_sync = crate::storage::parse_accounts(&storage_path, &accounts_before_sync)?;
-        let synced_accounts = sync_accounts(&storage_path, Some(0), &mut accounts_before_sync).await?;
-        let accounts_after_sync = crate::storage::with_adapter(&storage_path, |storage| storage.get_all())?;
-        let mut accounts_after_sync = crate::storage::parse_accounts(&storage_path, &accounts_after_sync)?;
+        let mut accounts_before_sync = Vec::new();
+        for account_handle in accounts.read().await.values() {
+            accounts_before_sync.push(account_handle.read().await.clone());
+        }
+        let synced_accounts = sync_accounts(accounts.clone(), &storage_path, Some(0)).await?;
+        let accounts_after_sync = accounts.read().await;
 
         // compare accounts to check for balance changes and new messages
         for account_before_sync in &accounts_before_sync {
             let account_after_sync = accounts_after_sync
-                .iter_mut()
-                .find(|account| account.id() == account_before_sync.id())
-                .unwrap();
+                .iter()
```

Comment:
Be aware that this is an O(n^2) operation. If it can be replaced with a hashmap lookup that would likely result in wins on workloads with many accounts.

Reply:
nice catch, totally forgot to change it since it`s now a HashMap instead of a Vec